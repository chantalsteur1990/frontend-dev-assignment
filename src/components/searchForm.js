import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import SearchContainer from './search/search.container'
import Suggestions from './suggestions/suggestions';
import * as searchFormActions from '../actions/searchFormActions';

export class SearchForm extends Component {
  constructor(props) {
    super(props);

    this.handleSuggestions = this.handleSuggestions.bind(this);
    this.handleSearch = this.handleSearch.bind(this);
    this.handleQueryChange = this.handleQueryChange.bind(this);

    this.state = {
      query: '',
      loading: false
    }
  }

  /**
   * Dispatch suggestionsRequest with
   * given query
   * 
   * @return {Void}
   */
  handleSuggestions = () => {
    const { actions: { suggestionsRequest } } = this.props;

    if (this.state.query.length > 1) {
      suggestionsRequest(this.state.query);
    }
  };

  /**
   * Dispatch searchRequest action
   * with value from local state
   * 
   * @return {Void}
   */
  handleSearch = () => {
    const { actions: { searchRequest } } = this.props;
    searchRequest(this.state.query);
  }

  /**
   * Set query in local state
   * 
   * @param {String} query
   * 
   * @return {Void}
   */
  handleQueryChange = (query) => {
    this.setState({
      ...this.state,
      query
    });
  }

  render() {
    const { query } = this.state;
    const { suggestions } = this.props;
    const showSuggestions = query.length > 1;
    return (
      <div>
        <SearchContainer
          handleSuggestions={this.handleSuggestions}
          handleSearch={this.handleSearch}
          handleQueryChange={this.handleQueryChange}
          value={query} />

          { showSuggestions &&
            <Suggestions
              suggestions={suggestions}
              query={query}
              handleSelect={this.handleQueryChange} />
          }
      </div>
    )
  }
};

const mapStateToProps = state => ({
  suggestions: state.searchForm.suggestions,
  loading: state.searchForm.loading
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(searchFormActions, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(SearchForm);