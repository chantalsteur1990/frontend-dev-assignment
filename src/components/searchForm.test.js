import React from 'react';
import { shallow, configure } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import { SearchForm } from './searchForm';

configure({ adapter: new Adapter() });

describe('SearchForm', () => {
  it('should render (snapshot)', () => {
    const component = shallow(<SearchForm />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should render with suggestions (snapshot)', () => {
    const propsFromState = {
      suggestions: [{
        searchterm: 'trui',
        nrResults: 123
      }]
    };
    const component = shallow(<SearchForm {...propsFromState} />);
    component.instance().handleQueryChange('tr');
    expect(shallowToJson(component)).toMatchSnapshot();
  });
  
  it('should only dispatch suggestionsRequest action when query lenght is > 1', () => {
    const propsFromState = {
      actions: {
        suggestionsRequest: jest.fn()
      }
    };
    const component = shallow(<SearchForm {...propsFromState} />);
    component.instance().handleSuggestions();
    expect(propsFromState.actions.suggestionsRequest).not.toHaveBeenCalled();
    component.instance().handleQueryChange('tr');
    component.instance().handleSuggestions();
    expect(propsFromState.actions.suggestionsRequest).toHaveBeenCalled();
  });

  it('should dispatch searchRequest action', () => {
    const propsFromState = {
      actions: {
        searchRequest: jest.fn()
      }
    };
    const component = shallow(<SearchForm {...propsFromState} />);
    component.instance().handleSearch();
    expect(propsFromState.actions.searchRequest).toHaveBeenCalled();
  });

  it('should have mutated the query in the state', () => {
    const component = shallow(<SearchForm />);
    expect(component.instance().state.query).toEqual('');
    component.instance().handleQueryChange('tr');
    expect(component.instance().state.query).toEqual('tr');
  });
});
