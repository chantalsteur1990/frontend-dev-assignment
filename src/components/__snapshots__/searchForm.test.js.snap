// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm should render (snapshot) 1`] = `
<div>
  <SearchContainer
    ariaLabelClear="annuleren"
    ariaLabelInput="zoeken"
    ariaLabelSubmit="zoeken"
    handleQueryChange={[Function]}
    handleSearch={[Function]}
    handleSuggestions={[Function]}
    name="search"
    placeholder="Zoeken"
    value=""
  />
</div>
`;

exports[`SearchForm should render with suggestions (snapshot) 1`] = `
<div>
  <SearchContainer
    ariaLabelClear="annuleren"
    ariaLabelInput="zoeken"
    ariaLabelSubmit="zoeken"
    handleQueryChange={[Function]}
    handleSearch={[Function]}
    handleSuggestions={[Function]}
    name="search"
    placeholder="Zoeken"
    value="tr"
  />
  <Suggestions
    handleSelect={[Function]}
    query="tr"
    suggestions={
      Array [
        Object {
          "nrResults": 123,
          "searchterm": "trui",
        },
      ]
    }
  />
</div>
`;
